### ISSUE TITLE
Small questions regarding compatibility with Geoopt optimizers

### ISSUE THREAD

**Issue by ncourty** [maxvanspengler/hyperbolic_learning_library#31]
Hello ! Thanks for the great work. We really look forward to test your library on our applications. I have a small question, which is related to a potential compatibility with geoopt. Let's say I have a loss function with parameters defined on other manifolds than the ones that are handled in your library. Do you believe that, with minimal adaptation, we can train the full pipeline (backbone from Hypll + loss defined wrt. geoopt) with an optimizer (RADAM or RSGD) from geoopt ?

**Comment by maxvanspengler** [comment]
Hi! Thank you for your interest in our library!

Regarding your question, if you were to use our ManifoldParameter class for this, then I think it would not work, as these shouldn't get recognized by the optimizers from geoopt. There would also probably be some other issues with the methods on the ManifoldParameter object itself, as these often assume the Manifold attribute to be structured as in our library. Lastly, the layers in our hypll.nn module expect the supplied manifold to have our structure as well, so this might also lead to some issues.

What manifold would you like to use from the geoopt library? We intend to add other manifolds such as the hyperboloid to our library in the future. So we can try to add the manifold you want sometime soon. The operations required for each of the layers have not yet been completely derived for some of the manifolds, so we decided to start with the Poincaré ball for now and add other manifolds later.

**Comment by ncourty** [comment]
Thanks for your quick reply ! And somehow I was expecting your answer. The types of manifold we wanted to use are hyperspheres, or SPDmatrix cones. In the end, I am wondering wether it could be of interest for you to support geoopt as an optional backend for the manifold data tensors. I read from your paper that your implementations of hyperbolic operations are optimized compared to geoopt ones (maybe explicit gradients ?), but maybe letting the choice of this backend to an end user might be more versatile. I am pretty conscious that this design choice might be very specific to my case, so maybe wait for more feedbacks from the community to see if it is worth thinking about it (if it appears to be the case, I would be glad to help).

**Comment by maxvanspengler** [comment]
We initially considered building our library upon geoopt, but we ran into several issues with their ManifoldTensor class that would be a bit problematic for our case. The largest issue was that most of their operations involving ManifoldTensors would return a torch.Tensor instead, so they drop the information regarding the manifold. We have designed our operations to return a ManifoldTensor instead (when appropriate), containing all the relevant information regarding the data. This design allows us to make the layers inside the hypll.nn module manifold agnostic, while also making debugging easier on our users. 

The drawback of our approach is that it makes compatibility with geoopt difficult. I haven't been able to think of a satisfactory solution to this problem yet. I think the easier solution would be to simply implement the manifolds within our library as well. Either way, it is certainly something to keep in mind as we continue and your help would be greatly appreciated when we make a decision on how to approach this, so thank you for raising this issue! I'll leave it open for further discussion.

**Comment by ncourty** [comment]
Thank you for the explanation and the discussion !

**Comment by ez2rok** [comment]
Insightful discussion. One other thought: have you considered making your library compatible with [Geomstats](https://geomstats.github.io/)?

**Comment by maxvanspengler** [comment]
I am not familiar with the Geomstats package, but it looks very interesting! I'm not sure how the two libraries would interact, but if you have some examples I'd be happy to take a look.

Based on my first glance it appears that their library would have a somewhat similar function to scikit-learn, but for differentiable manifolds (please correct me if I'm wrong). If that's the case, then I think it would be best if the relation between Geomstats and HypLL is similar to the relation between PyTorch and scikit-learn, where they do not really interact much but are used for slightly different tasks and where scikit-learn is sometimes used to evaluate PyTorch models.
